Sources/Schema.swift

```diff
@@ -100,7 +101,27 @@ extension SchemaBuilder {
         }
 
         let name = components[2]
+
+        if let cachedSchema = ReferenceResolver.cachedSchemaReferences[name] {
+            return Structure(name: name, structure: cachedSchema)
+        }
+        if let resolving = ReferenceResolver.resolvingReferences[name], resolving {
+            return Structure(name: name, structure: .unresolvedReference(pointer.path))
+        }
+        ReferenceResolver.resolvingReferences[name] = true
         let schema = try builder.build(swagger)
+        ReferenceResolver.cachedSchemaReferences[name] = schema
+        ReferenceResolver.resolvingReferences[name] = nil
         return Structure(name: name, structure: schema)
     }
 }
+
+enum ReferenceResolver {
```

Comment:
Instead of exposing the dictionaries themselves, how about we make a nicer public API for ReferenceResolver that is extensible to other types as well:
```
protocol ResolvableType: Builder {
    static var path: String { get }
}

private let kSchemaBuilderReferenceResolver = ReferenceResolver<SchemaBuilder>()
extension SchemaBuilder: ResolvableType {
    static var path: String { return "definitions" }
    static var resolver: ReferenceResolver<SchemaBuilder> { return kSchemaBuilderReferenceResolver }
}

private let kParameterBuilderReferenceResolver = ReferenceResolver<ParameterBuilder>()
extension ParameterBuilder: ResolvableType {
    static var path: String { return "parameters" }
    static var resolver: ReferenceResolver<ParameterBuilder> { return kParameterBuilderReferenceResolver }
}

private let kResponseBuilderReferenceResolver = ReferenceResolver<ResponseBuilder>()
extension ResponseBuilder: ResolvableType {
    static var path: String { return "responses" }
    static var resolver: ReferenceResolver<ResponseBuilder> { return kResponseBuilderReferenceResolver }
}


class ReferenceResolver<T: ResolvableType> {
    enum ResolverError: Error {
        case invalidContext
        case invalidReference
        case invalidPath
        case unsupportedReference
        case unresolvedReference
        case circularReference
    }

    private struct Context<C> {
        var cachedReferences = [String: C]()
        var resolvingReferences = [String: Bool]()
    }

    private var context: Context<T>?

    func setup() {
        self.context = Context()
    }

    func resolve(_ swagger: SwaggerBuilder, pointer: Pointer<T>) throws -> Structure<T.Building> {
        guard let context = self.context else {
            throw ResolverError.invalidContext
        }

        let components = pointer.path.components(separatedBy: "/")
        guard components.count == 3 && components[0] == "#" && components[1] == T.path else {
            throw ResolverError.invalidPath
        }

        let name = components[2]
        let builder: T?
        switch T.path {
        case "definitions": builder = swagger.definitions[name]
        case "parameters": builder = swagger.parameters[name]
        case "responses": builder = swagger.responses[name]
        default: throw ResolverError.unsupportedReference
        }

        guard let builder = builder else {
            throw ResolverError.unresolvedReference
        }

        if let cached = context.cachedReferences[name] {
            return return Structure(name: name, structure: cached)
        }

        if let resolving = context.resolvingReferences[name], resolving {
            throw ResolverError.circularReference
        }

        context.resolvingReferences[name] = true
        let resolved = try builder.build(swagger)
        context.cachedReferences[name] = resolved
        context.resolvingReferences[name] = nil
        return return Structure(name: name, structure: resolved)
    }

    func teardown() {
        self.context  = nil
    }
}
```

It should also be moved to its own file.